File: luna-sdk-react-native/src/api/PriorityFeeApi.ts
/**
 * PriorityFeeApi - Priority Fee Estimation
 * 
 * Get optimal priority fees for fast transaction confirmation
 */

import type { LunaHeliusClient } from '../LunaHeliusClient';
import type { RpcResponse, PriorityFeeEstimate } from '../types';

export interface PriorityFeeOptions {
  accountKeys?: string[];
  includeAllPriorityFeeLevels?: boolean;
  lookbackSlots?: number;
  recommended?: boolean;
}

export class PriorityFeeApi {
  constructor(private readonly client: LunaHeliusClient) {}

  /** Get priority fee estimate */
  async getPriorityFeeEstimate(
    options?: PriorityFeeOptions
  ): Promise<RpcResponse<{ priorityFeeLevels: PriorityFeeEstimate }>> {
    const params: any = {};
    
    if (options?.accountKeys) {
      params.accountKeys = options.accountKeys;
    }
    if (options?.includeAllPriorityFeeLevels !== undefined) {
      params.options = {
        ...params.options,
        includeAllPriorityFeeLevels: options.includeAllPriorityFeeLevels,
      };
    }
    if (options?.lookbackSlots !== undefined) {
      params.options = {
        ...params.options,
        lookbackSlots: options.lookbackSlots,
      };
    }
    if (options?.recommended !== undefined) {
      params.options = {
        ...params.options,
        recommended: options.recommended,
      };
    }

    return this.client.rpcCall('getPriorityFeeEstimate', params);
  }

  /** Get recent priority fees from the network */
  async getRecentPriorityFees(
    accounts?: string[]
  ): Promise<RpcResponse<any[]>> {
    const params = accounts ? [accounts] : [];
    return this.client.rpcCall('getRecentPrioritizationFees', params);
  }

  /** Calculate optimal fee for a transaction */
  async calculateOptimalFee(
    serializedTransaction: string,
    options?: {
      priorityLevel?: 'min' | 'low' | 'medium' | 'high' | 'veryHigh' | 'unsafeMax';
    }
  ): Promise<RpcResponse<{ fee: number; priorityLevel: string }>> {
    // Simulate to get accounts involved
    const simulation = await this.client.rpcCall('simulateTransaction', [
      serializedTransaction,
      { encoding: 'base64' },
    ]);

    if (simulation.error) {
      return { result: null, error: simulation.error };
    }

    // Get priority fee estimate for those accounts
    const accounts = simulation.result?.value?.accounts?.map((a: any) => a?.owner).filter(Boolean) || [];
    
    const feeEstimate = await this.getPriorityFeeEstimate({
      accountKeys: accounts,
      includeAllPriorityFeeLevels: true,
    });

    if (feeEstimate.error || !feeEstimate.result) {
      return { result: null, error: feeEstimate.error };
    }

    const levels = feeEstimate.result.priorityFeeLevels;
    const level = options?.priorityLevel ?? 'medium';
    
    const fee = levels[level as keyof PriorityFeeEstimate] ?? levels.medium;

    return {
      result: { fee, priorityLevel: level },
      error: null,
    };
  }
}

File: luna-priority/src/main/kotlin/xyz/selenus/luna/priority/PriorityFeeApi.kt
package xyz.selenus.luna.priority

import kotlinx.coroutines.delay
import kotlinx.serialization.json.*
import xyz.selenus.luna.LunaHeliusClient
import xyz.selenus.luna.RpcError
import xyz.selenus.luna.RpcResponse

/**
 * Priority fee estimation API.  Use this to estimate network fees for a transaction
 * given a desired priority level.
 */
class PriorityFeeApi internal constructor(private val client: LunaHeliusClient) {
    /**
     * Estimate the fee per compute unit needed to achieve a certain priority level.
     * @param transaction The base58 or base64 encoded transaction string.
     * @param accountKeys A list of account keys involved in the transaction (alternative to passing transaction).
     * @param priorityLevel One of "Min", "Low", "Medium", "High", "VeryHigh", "UnsafeMax", or "Default".
     * @param includeAllPriorityFeeLevels If true, returns estimates for all priority levels.
     * @param lookbackSlots Number of slots to look back for estimation.
     */
    suspend fun getPriorityFeeEstimate(
        transaction: String? = null,
        accountKeys: List<String>? = null,
        priorityLevel: String? = null,
        includeAllPriorityFeeLevels: Boolean? = null,
        lookbackSlots: Int? = null,
        recommended: Boolean? = null
    ): RpcResponse<JsonElement> {
        val options = buildJsonObject {
            priorityLevel?.let { put("priorityLevel", it) }
            includeAllPriorityFeeLevels?.let { put("includeAllPriorityFeeLevels", it) }
            lookbackSlots?.let { put("lookbackSlots", it) }
            recommended?.let { put("recommended", it) }
        }

        val paramsObj = buildJsonObject {
            transaction?.let { put("transaction", it) }
            accountKeys?.let { keys ->
                put("accountKeys", JsonArray(keys.map { JsonPrimitive(it) }))
            }
            if (options.isNotEmpty()) {
                put("options", options)
            }
        }
        
        // The RPC expects an array containing the parameters object
        val params = buildJsonArray { add(paramsObj) }
        return client.rpcCall("getPriorityFeeEstimate", params)
    }
}

/**
 * Access the PriorityFeeApi namespace from a [LunaHeliusClient].
 *
 * Import this extension to enable the `client.priority` style:
 * ```
 * import xyz.selenus.luna.priority.priority
 * client.priority.<method>()
 * ```
 */
val LunaHeliusClient.priority: PriorityFeeApi
    get() = PriorityFeeApi(this)
